crates/common/fork_choice/src/handlers.rs

```diff
@@ -0,0 +1,205 @@
+    }
+
+    verify_blob_kzg_proof_batch(&blobs, blob_kzg_commitments, &proofs)?;
+    Ok(true)
+}
+
+pub fn get_ancestor(store: &Store, root: B256, slot: u64) -> B256 {
+    if let Some(block) = store.blocks.get(&root) {
+        if block.slot > slot {
+            return get_ancestor(store, block.parent_root, slot);
+        }
+    }
+    root
+}
+
+/// Compute the checkpoint block for epoch ``epoch`` in the chain of block ``root``
+pub fn get_checkpoint_block(store: &Store, root: B256, epoch: u64) -> B256 {
+    let epoch_first_slot = compute_start_slot_at_epoch(epoch);
+    get_ancestor(store, root, epoch_first_slot)
+}
+
+/// Update checkpoints in store if necessary
+pub fn update_checkpoints(
+    store: &mut Store,
+    justified_checkpoint: Checkpoint,
+    finalized_checkpoint: Checkpoint,
+) {
+    // Update justified checkpoint
+    if justified_checkpoint.epoch > store.justified_checkpoint.epoch {
+        store.justified_checkpoint = justified_checkpoint
+    }
+    // Update finalized checkpoint
+    if finalized_checkpoint.epoch > store.finalized_checkpoint.epoch {
+        store.finalized_checkpoint = finalized_checkpoint
+    }
+}
+
+/// Update unrealized checkpoints in store if necessary
+pub fn update_unrealized_checkpoints(
+    store: &mut Store,
+    unrealized_justified_checkpoint: Checkpoint,
+    unrealized_finalized_checkpoint: Checkpoint,
+) {
+    // Update unrealized justified checkpoint
+    if unrealized_justified_checkpoint.epoch > store.unrealized_justified_checkpoint.epoch {
+        store.unrealized_justified_checkpoint = unrealized_justified_checkpoint
+    }
+    // Update unrealized finalized checkpoint
+    if unrealized_finalized_checkpoint.epoch > store.unrealized_finalized_checkpoint.epoch {
+        store.unrealized_finalized_checkpoint = unrealized_finalized_checkpoint
+    }
+}
+
+pub fn get_current_store_epoch(store: &Store) -> u64 {
+    compute_epoch_at_slot(store.get_current_slot())
+}
+
+pub fn compute_pulled_up_tip(store: &mut Store, block_root: B256) -> anyhow::Result<()> {
+    let mut state = store.block_states[&block_root].clone();
+    // Pull up the post-state of the block to the next epoch boundary
+    state.process_justification_and_finalization()?;
+
+    store
+        .unrealized_justifications
+        .insert(block_root, state.current_justified_checkpoint);
+    update_unrealized_checkpoints(
+        store,
+        state.current_justified_checkpoint,
+        state.finalized_checkpoint,
+    );
+
+    // If the block is from a prior epoch, apply the realized values
+    let block_epoch = compute_epoch_at_slot(store.blocks[&block_root].slot);
+    let current_epoch = get_current_store_epoch(store);
+    if block_epoch < current_epoch {
+        store.update_checkpoints(
+            state.current_justified_checkpoint,
+            state.finalized_checkpoint,
+        );
+    }
+
+    Ok(())
+}
+
+/// Run ``on_block`` upon receiving a new block.
+pub async fn on_block(
```

Comment:
I think it's better to all functions other than `on_block` here can be moved to store.rs? We can place `on_tick`, `on_attestation` and `on_attester_slashing` in handler.rs.